bindings/python/pyiree/test/tree_utils.py

```diff
@@ -0,0 +1,195 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""A minimal implementation of XLA's pytree flattening and unflattening."""
+
+# TODO(#4131) python>=3.7: Use postponed type annotations.
+
+import copy
+from typing import Any, Callable, Dict, List, Mapping, Sequence, Tuple, Union
+
+__all__ = [
+    "is_leaf",
+    "tree_flatten",
+    "tree_map",
+    "tree_unflatten",
+]
+
+Leaf = Any
+Branch = Union[Sequence["Tree"], Mapping[Any, "Tree"]]
+Tree = Union[Leaf, Branch]  # pytype: disable=not-supported-yet
+
+TreeDefLeaf = None
+TreeDefBranch = Union[List["TreeDef"], Tuple["TreeDef"], Dict[Any, "TreeDef"]]
+TreeDef = Union[TreeDefLeaf, TreeDefBranch]  # pytype: disable=not-supported-yet
+
+
+def is_leaf(tree: Tree) -> bool:
+  return not isinstance(tree, (Sequence, Mapping))
+
+
+def _set_element(tree: TreeDefBranch, index: Any, value: Tree) -> Branch:
```

Comment:
Ok so index is either a number indicating a subscript into an array/list/tuple or it's a dict key? Would it be better to have some kind of checks around that or is error checking not pythonic :stuck_out_tongue: